presto-main/src/main/java/com/facebook/presto/util/Failures.java

```diff
@@ -71,28 +89,34 @@ public static ExecutionFailureInfo toFailure(Throwable failure)
             remoteHost = ((PrestoTransportException) failure).getRemoteHost();
         }
 
+        Throwable cause = failure.getCause();
+        if (cause != null && seenFailures.contains(cause)) {
+            cause = null;
+        }
+        else {
+            seenFailures.add(cause);
```

Comment:
`add` returns a boolean that tells whether the new element already exists. You can use that to simplify that if/else a bit.